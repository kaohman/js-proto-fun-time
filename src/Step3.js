import React from 'react';

function Step3(props) {
  return (
    <div className='step-containers'>
      <div>
        <h3>Step 3: Let's Solve The Problem</h3>
        <p>Here's the setup:</p>
        <textarea rows='15' cols='40'></textarea>
      </div>
      <button>Check Answer</button>
    </div>
  )
}

export default Step3;