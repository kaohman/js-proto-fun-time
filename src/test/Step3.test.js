import React from 'react';
import ReactDOM from 'react-dom';
import Step3 from '../Step3';
import { shallow } from 'enzyme';

const incrementStepMock = jest.fn();
const correctAnswerMock = 3;
const inputMock = [1, 2, 3];
const gameLengthMock = 31;
const currentStepMock = 3;
const parsedInputMock = '[1, 2, 3]';
const questionCountMock = 4;
const newCodeMock = `function solveProblem(arg) { return 3 }
  
  var result = solveProblem(inputData);`;
let eventMock = { preventDefault: () => { } };

describe('Step3', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <Step3 
        incrementStep={incrementStepMock}
        correctAnswer={correctAnswerMock}
        input={inputMock}
        renderInput={parsedInputMock}
        questionCount={questionCountMock}
        gameLength={gameLengthMock}
        currentStep={currentStepMock}
      />
    );
  });

  it('should match the snapshot with all data passed in', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have a default state', () => {
    expect(wrapper.state()).toEqual({ 
      userCode: '',
      correctResponse: ''
     });
  });

  it('should get answer input as the user types in the code editor', () => {
    expect(wrapper.state('userCode')).toEqual('');
    wrapper.find('.code-editor').simulate('change', { getValue: () => { return newCodeMock } });
    expect(wrapper.state('userCode')).toEqual(`function solveProblem(arg) { return 3 }
  
  var result = solveProblem(inputData);`);
  });

  it('should be able to tell the user if their answer is correct', () => {
    wrapper.setState({ userCode: newCodeMock });
    wrapper.find('#check-answer-button-3').simulate('click', eventMock);
    expect(wrapper.state('correctResponse')).toEqual('Correct! Click Next Problem to continue to the next problem.');
  });
  
  it('should be able to tell the user if their answer is incorrect', () => {
    wrapper.setState({
      userCode: `function solveProblem(arg) { return 1 }
  
  var result = solveProblem(inputData);`});
    wrapper.find('#check-answer-button-3').simulate('click', eventMock);
    expect(wrapper.state('correctResponse')).toEqual('Sorry that is incorrect, please try again.');
  });

  it('should allow the user to click the Next Problem button if their answer was correct', () => {
    wrapper.setState({ correctResponse: 'Correct! Click Next Problem to continue to the next problem.' });
    wrapper.find('#next-step-button-3').simulate('click');
    expect(incrementStepMock).toBeCalled();
  });  

  it('should allow the user to click the New Game button if their answer was correct and they have completed the final problem', () => {
    wrapper.setState({ correctResponse: 'Correct, you completed all the problems! Click New Game.'});
    wrapper.find('#next-step-button-3').simulate('click');
    expect(incrementStepMock).toBeCalled();
  });  
  
});