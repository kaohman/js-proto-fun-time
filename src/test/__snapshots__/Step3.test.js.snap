// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step3 should match the snapshot with all data passed in 1`] = `
<div
  className="step-containers"
  id="step-3-container"
>
  <div>
    <h3>
      Step 3: Let's Solve The Problem
    </h3>
    <p>
      Here's the setup:
    </p>
    <UnControlled
      className="code-editor"
      onChange={[Function]}
      options={
        Object {
          "lineNumbers": true,
          "lineWrapping": true,
          "mode": "javascript",
          "theme": "material",
        }
      }
      value="  var inputData = [1, 2, 3];
      
  function solveProblem(arg) {
    // Apply the prototype to your input here
    // Remember to return the result.
  }
  
  var result = solveProblem(inputData);"
    />
    <h4 />
  </div>
  <button
    id="check-answer-button-3"
    onClick={[Function]}
  >
    Check Answer
  </button>
</div>
`;
